sdk/core/core-amqp/test/lock.spec.ts

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortController, AbortError } from "@azure/abort-controller";
+import * as chai from "chai";
+import { OperationTimeoutError } from "rhea-promise";
+import { delay } from "../src";
+const should = chai.should();
+
+import { CancellableAsyncLock, CancellableAsyncLockImpl } from "../src/util/lock";
+import { settleAllTasks } from "./utils/utils";
+
+describe("CancellableAsyncLock", function() {
+  const TEST_FAILURE = "Test failure";
+
+  describe(".acquire", function() {
+    let lock: CancellableAsyncLock;
+    beforeEach("create lock", () => {
+      lock = new CancellableAsyncLockImpl();
+    });
+
+    it("forwards values from task", async () => {
+      const expectedValues = ["foo", "bar", 3.14159, new Date(), {}, null];
+
+      const tasks: Promise<any>[] = [];
+      for (const val of expectedValues) {
+        tasks.push(lock.acquire("lock", async () => val));
+      }
+
+      const results = await Promise.all(tasks);
+      results.should.deep.equal(expectedValues, "Unexpected value returned from tasks.");
+    });
+
+    it("forwards error from task", async () => {
+      try {
+        await lock.acquire("lock", async () => {
+          throw new Error("I break things!");
+        });
+        throw new Error(TEST_FAILURE);
+      } catch (err) {
+        should.equal(err.message, "I break things!");
+      }
+    });
+
+    it("works using single key", async () => {
+      const taskCount = 10;
+      const tasks: Promise<number>[] = [];
+      for (let i = 0; i < taskCount; i++) {
+        tasks.push(
+          lock.acquire("lock", async () => {
+            // Add a delay such that later tasks would resolve
+            // faster than early tasks if they all ran at the
+            // same time.
+            await delay(taskCount - i);
+            return i;
+          })
+        );
+      }
+
+      // verify order
+      for (let i = 0; i < taskCount; i++) {
+        const result = await Promise.race(tasks);
+        should.equal(result, i, "Tasks ran out of order.");
+        // Since tasks should be completed in order, remove head task.
+        tasks.shift();
+      }
+      should.equal(tasks.length, 0, "There are still tasks pending.");
+    });
+
+    it("keys are isolated", async () => {
+      /*
+        We enqueue 2 tasks on key "1", and 2 tasks on key "2".
+        Each task has a delay so that it yields control to the event loop
+        before resolving.
+
+        The 1st task on key "1" will resolve first.
+        The 1st task on key "2" will resolve second because it does not
+        need to wait for tasks on key "1" to resolve.
+        The 2nd task on key "1" will resolve third. It must wait for
+        the 1st tak on key "1" to resolve.
+        The 2nd task on key "2" will resolve last. It must wait for
+        the 1st task on key "2" to resolve.
+      */
+      const tasks = [
```

Comment:
I'm not sure how sinon.clock() helps with that. I'm basically calling setTimeout(0) just so that the order things are queued up in for the event loop is known.

Reply:
Yeah, that's fine, it was more a question than a demand :)